src/NServiceBus.Core/Persistence/InMemory/Outbox/InMemoryOutboxStorage.cs

```diff
@@ -48,24 +47,21 @@ public Task SetAsDispatched(string messageId, ContextBag context)
                 return TaskEx.CompletedTask;
             }
 
-            storedMessage.TransportOperations.Clear();
-            storedMessage.Dispatched = true;
-
+            storedMessage.MarkAsDispatched();
             return TaskEx.CompletedTask;
         }
 
         public void RemoveEntriesOlderThan(DateTime dateTime)
         {
-            var entriesToRemove = storage
-                .Where(e => e.Value.Dispatched && e.Value.StoredAt < dateTime)
-                .Select(e => e.Key)
-                .ToList();
-
-            foreach (var entry in entriesToRemove)
+            foreach (var entry in storage.ToArray())
```

Comment:
That's safe to do with `ConcurrentDictionary`. 

https://msdn.microsoft.com/en-us/library/dd287131(v=vs.110).aspx

And since this is trying to purge old values, it shouldn't matter that it's not a moment-in-time snapshot either.